### PR TITLE
ToListAndFree usage

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Usage of PR #25563

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Parser/ParseDeclarationStatement.vb [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+Imports System.Runtime.InteropServices
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.PooledObjects
+Imports Microsoft.CodeAnalysis.Syntax.InternalSyntax
+Imports Microsoft.CodeAnalysis.Text
+Imports CoreInternalSyntax = Microsoft.CodeAnalysis.Syntax.InternalSyntax
+Imports InternalSyntaxFactory = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SyntaxFactory
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Parser/ParseDeclarationStatement.vb]: why was all this code created?

**Comment by CyrusNajmabadi** [comment]
Please do not make unrelated refactorings when doing these PRs.  It makes it enormously difficult to validate and verify.

File: src/Compilers/VisualBasic/Portable/Parser/ParseDeclarationStatement.vb (hunk 1 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseDeclarationStatement.vb]: I separated out some of the parser methods into their own files. I also moved, some the of the xml parser  from `Parser.vb` is not in the `ParseXML.vb` file.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi Revert the new files, rebased and sqaushed.

**Comment by CyrusNajmabadi** [comment]
Just my 2c, but you've made this PR much harder than necessary by interleaving unrelated changes with your change to use the new API.  In its current form, it's not something i coudl review.

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 2]
```diff
@@ -378,12 +380,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             ' grab the part that doesn't contain the preceding and trailing trivia.
 
             Dim builder = PooledStringBuilder.GetInstance()
-            Dim writer As New IO.StringWriter(builder)
-
-            firstToken.WriteTo(writer)
-            secondToken.WriteTo(writer)
-            thirdToken.WriteTo(writer)
-
+            Using writer As New IO.StringWriter(builder)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Leaving change in as the `writer` wasn't being disposed.

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 3]
```diff
@@ -352,15 +353,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
         Private Shared Function MergeTokenText(firstToken As SyntaxToken, secondToken As SyntaxToken) As String
 
             ' grab the part that doesn't contain the preceding and trailing trivia.
 
             Dim builder = PooledStringBuilder.GetInstance()
-            Dim writer As New IO.StringWriter(builder)
-
-            firstToken.WriteTo(writer)
-            secondToken.WriteTo(writer)
+            Using writer As New IO.StringWriter(builder)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Leaving change in as the `writer` wasn't being disposed.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi Is this better for reviewing?

**Comment by CyrusNajmabadi** [comment]
yes.  Much better.  thanks!

**Comment by AdamSpeight2008** [comment]
@dotnet-bot Retest this please.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot Retest this please.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot test this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot test this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot test this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest windows_debug_spanish_unit32_prtest please

File: src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb [hunk 4]
```diff
@@ -711,12 +702,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
-            Dim joinVarList = New CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList(Of CollectionRangeVariableSyntax)(joinVariable)
-            Dim moreJoins = moreJoinsBuilder.ToList()
-            _pool.Free(moreJoinsBuilder)
+            Dim joinVarList = _pool.AllocateSeparated(Of CollectionRangeVariableSyntax)
+            joinVarList.Add(joinVariable)
+            Dim moreJoins = _pool.ToListAndFree(moreJoinsBuilder)
 
             If groupKw Is Nothing Then
-                Return SyntaxFactory.SimpleJoinClause(joinKw, joinVarList, moreJoins, onKw, Predicate)
+                Return SyntaxFactory.SimpleJoinClause(joinKw, _pool.ToListAndFree(joinVarList), moreJoins, onKw, Predicate)
```
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb]: the list seems can be re-used.

File: src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb [hunk 5]
```diff
@@ -2212,6 +2192,8 @@ TryResync:
             Else
                 Return node
             End If
+
+            Return node
```
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb]: why is this here? is it reachable?

File: src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb [hunk 6]
```diff
@@ -2229,6 +2211,8 @@ TryResync:
...
```
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb]: same here and below

File: src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb (hunk 4 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb]: different lists `moreJoinsBuilder` and `joinVarList`

File: src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb (hunk 6 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb]: Well spotted I remove the last one,

File: src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb (hunk 5 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseXml.vb]: Well spotted I remove the last one,

File: src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb (hunk 4 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb]: @heejaechang  I've made them extension methods, so that it better matches meaning.

Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb]: Also they are on different function termination paths so can be reused.
